src/ir_passes/phantom_check.rs

```diff
@@ -0,0 +1,127 @@
+use itertools::Itertools;
+
+use crate::{
+    diagnostics,
+    errors::Error,
+    ir::{self, Ctx},
+    ir_visitor::{Action, Visitor, VisitorData},
+    utils::GPosIdx,
+};
+use std::collections::HashSet;
+
+#[derive(Default)]
+/// Checks if a user-level phantom events are valid.
+/// Phantom events are valid iff:
+/// 1. The component doesn't share any instances
+/// 2. The component doesn't use an subcomponents that need to use the
+///    corresponding event in their interface, i.e., the uses of the event are all phantom
+pub struct PhantomCheck {
+    instance_used: HashSet<ir::InstIdx>,
+    prev_invoke: Option<ir::InvIdx>,
+    phantom_events: Vec<ir::EventIdx>,
+    diag: diagnostics::Diagnostics,
+    diag_count: u32,
```

Comment:
The `Diagnostics` struct returns a count of the errors when you call `report_all` on it. The general idea with it is you build up all the errors in a pass and then report them together. By default, if a pass implements `Default`, we [create a new instance of the struct](https://www.irisa.fr/alf/downloads/ADA/p127-josipovic.pdf) every time we visit a new component so the `diag` object is not being shared here. To share information, you have to manually override `clear_data`.

We probably need to do some restructuring to make sure we report *all* phantom check problems in all components using this pass. Maybe a test where we have two components with problems and at least one of them has two phantom check violations would be useful.